flink/src/main/java/org/apache/iceberg/flink/data/FlinkSchemaVisitor.java

```diff
@@ -29,42 +29,43 @@
 import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 import org.apache.iceberg.types.Type;
 import org.apache.iceberg.types.Types;
+import org.apache.orc.TypeDescription;
 
 abstract class FlinkSchemaVisitor<T> {
 
-  static <T> T visit(RowType flinkType, Schema schema, FlinkSchemaVisitor<T> visitor) {
-    return visit(flinkType, schema.asStruct(), visitor);
+  static <T> T visit(RowType flinkType, Schema schema, TypeDescription typeDesc, FlinkSchemaVisitor<T> visitor) {
```

Comment:
I think it would be cleaner to add the `beforeField`/`afterField` callbacks instead of the `typeDesc`.